language/move-lang/src/expansion/translate.rs

```diff
@@ -661,6 +670,49 @@ fn struct_fields(
     E::StructFields::Defined(field_map)
 }
 
+//**************************************************************************************************
+// Friends
+//**************************************************************************************************
+
+fn friend(context: &mut Context, friends: &mut UniqueMap<ModuleIdent, Loc>, pfriend: P::Friend) {
+    match friend_(context, pfriend) {
+        Some((mident, loc)) => {
+            if let Some(old_friend_loc) = friends.get(&mident) {
+                let msg = format!(
+                    "Duplicate friend declaration '{}'. Friend declarations in a module must be unique",
+                    mident
+                );
+                context.error(vec![
+                    (loc, msg),
+                    (*old_friend_loc, "Previously declared here".into()),
+                ]);
+            }
+            if let Err(_old_loc) = friends.add(mident, loc) {
+                assert!(context.has_errors());
+            };
```

Comment:
I would just do the `friends.add`. I know you will have to clone the `mident` but...
- I hope to intern all of the strings some day so that should be cheap
- I plan on updating the `add` api to remove the clone anyway

Reply:
Meant to make that last one a review comment, sorry for the spam. 

But in other words, I'd rather keep the style consistent even if it is "expensive" because it will make it easier to migrate everything later

Reply:
I totally buy this. But it is a little bit tricky in this case, as
```
fn add(&mut self, key: K, value: V) -> Result<(), K::Loc>
```
does not return `V` on error, and given the declaration of
```
friends: UniqueMap<ModuleIdent, Loc>
```
we do need the `V` here for the previously declared location of friend.

Reply:
Ah shit right. Yup, it is the wrong loc